python/ray/autoscaler/resource_demand_scheduler.py

```diff
@@ -0,0 +1,197 @@
+            instance_type_counts[instance_type] += 1
+
+        for node_id in nodes:
+            tags = self.provider.node_tags(node_id)
+            if TAG_RAY_INSTANCE_TYPE in tags:
+                instance_type = tags[TAG_RAY_INSTANCE_TYPE]
+                add_instance(instance_type)
+
+        for instance_type, count in pending_nodes.items():
+            for _ in range(count):
+                add_instance(instance_type)
+
+        return node_resources, instance_type_counts
+
+    def get_instances_to_launch(self, nodes: List[str],
+                                pending_nodes: Dict[str, int],
+                                resource_demands: List[dict]):
+        """Get a list of instance types that should be added to the cluster.
+
+        This method:
+            (1) calculates the resources present in the cluster.
+            (2) calculates the unfulfilled resource bundles.
+            (3) calculates which instances need to be launched to fulfill all
+                the bundle requests, subject to max_worker constraints.
+        """
+
+        if resource_demands is None:
+            logger.info("No resource demands")
+            return []
+
+        node_resources, instance_type_counts = self.calculate_node_resources(
+            nodes, pending_nodes)
+        logger.info("Cluster resources: {}".format(node_resources))
+        logger.info("Instance counts: {}".format(instance_type_counts))
+
+        unfulfilled = get_bin_pack_residual(node_resources, resource_demands)
+        logger.info("Resource demands: {}".format(resource_demands))
+        logger.info("Unfulfilled demands: {}".format(unfulfilled))
+
+        instances = get_instances_for(
+            self.instance_types, instance_type_counts,
+            self.max_workers - len(nodes), unfulfilled)
+        logger.info("Instance requests: {}".format(instances))
+        return instances
+
+
+# TODO(ekl) we could use ortools to minimize cost here instead.
+def get_instances_for(instance_types: Dict[str, dict],
+                      existing_instances: Dict[str, int], max_to_add: int,
+                      resources: List[dict]) -> List[Tuple[str, int]]:
+    instances_to_add = collections.defaultdict(int)
+    allocated_resources = []
+
+    while resources and sum(instances_to_add.values()) < max_to_add:
+        utilization_scores = []
+        for instance_type in instance_types:
+            if (existing_instances.get(
+                    instance_type, 0) + instances_to_add.get(instance_type, 0)
+                    >= instance_types[instance_type]["max_workers"]):
+                continue
+            node_resources = instance_types[instance_type]["resources"]
+            score = _utilization_score(node_resources, resources)
+            if score is not None:
+                utilization_scores.append((score, instance_type))
+
+        # Give up, no feasible node.
+        if not utilization_scores:
+            break
```

Comment:
How does the user find out about this case? Should we print an error message/throw an error?